### PR TITLE
private/michael/private/michael/imageperf

### PR DESCRIPTION
**Pull request by mmeeks**
- pure re-factor, creating Buffer class to wrap vector.
- Adapt Buffer class to avoid repeated compactions.

### CODE REVIEW

File: net/Buffer.hpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        }
+
+        _buffer.erase(_buffer.begin(), _buffer.begin() + _offset + len);
+        _offset = 0;
+        _size = _buffer.size() - _offset;
+    }
+    void append(const char *data, const int len)
+    {
+        _buffer.insert(_buffer.end(), data, data + len);
+        _size = _buffer.size() - _offset;
```
Comment (Ashod) [inline net/Buffer.hpp:60]: Or `_size += len;` which to me is more intuitive.

File: net/Buffer.hpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            len < _buffer.size() &&   // not a complete erase
+            _offset < 16384 * 64 &&   // do cleanup a Mb at a time or so:
+            _size > 512)              // early cleanup if what remains is small.
+        {
+            _offset += len;
+            _size -= len;
+            return;
+        }
+
+        _buffer.erase(_buffer.begin(), _buffer.begin() + _offset + len);
```
Comment (Ashod) [inline net/Buffer.hpp:53]: Shrinking should be as easy as:
```
        if (_buffer.size() >= 16384 * 64) // Or even 512kb.
            _buffer = std::vector<char>(_buffer.begin() + _offset + len, _buffer.end());
        else
            _buffer.erase(_buffer.begin(), _buffer.begin() + _offset + len);
```

File: net/Buffer.hpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    void eraseFirst(size_t len)
+    {
+        assert(len >= 0);
+        assert(_offset + len <= _buffer.size());
+
+        // avoid regular shuffling down larger chunks of data
+        if (_buffer.size() > 16384 && // lots of queued data
+            len < _buffer.size() &&   // not a complete erase
+            _offset < 16384 * 64 &&   // do cleanup a Mb at a time or so:
+            _size > 512)              // early cleanup if what remains is small.
```
Comment (Ashod) [inline net/Buffer.hpp:46]: Should really make these constants into `static constexpr` (private) members with descriptive names etc.

**Review by Ashod — COMMENTED** [review]
Looks good, runs great! Noticeable performance improvement with large images (and even with rapid text input) for sure.

Some notes and suggestions. A few are nitpicks, but hopefully useful stuff overall. (As an aside,  clang-format?)

Thanks!

File: net/Socket.hpp [hunk 4]
```diff
@@ -1153,17 +1152,18 @@ class StreamSocket : public Socket,
             ssize_t len;
             do
             {
-                // Writing more than we can absorb in the kernel causes SSL wastage.
-                len = writeData(&_outBuffer[0], std::min((int)_outBuffer.size(),
-                                                         getSendBufferSize()));
+                // Writing much more than we can absorb in the kernel causes wastage.
+                len = writeData(_outBuffer.getBlock(),
+                                std::min((int)_outBuffer.getBlockSize(),
```
Comment (Ashod) [inline net/Socket.hpp:1157]: Any reason not to use `_outBuffer.size()` here?

Comment (mmeeks) [reply net/Socket.hpp:1157]: Sure that's easy - we'll switch to having blocks/chunks under the hood that are not all in the same allocations in some star-trek future where we have time =)

**Pull request merged**